src/api.rs

```diff
@@ -683,8 +683,10 @@ pub(crate) struct ContextInner<T: Pixel> {
   keyframes: BTreeSet<u64>,
   /// A storage space for reordered frames.
   packet_data: Vec<u8>,
-  gop_output_frameno_start: u64,
-  pub(crate) gop_input_frameno_start: u64,
+  /// Maps `output_frameno` to `gop_output_frameno_start`.
+  gop_output_frameno_start: BTreeMap<u64, u64>,
+  /// Maps `output_frameno` to `gop_input_frameno_start`.
+  pub(crate) gop_input_frameno_start: BTreeMap<u64, u64>,
```

Comment:
This structure is technically redundant with the keyframe BTreeSet, isn't it?

Reply:
There's some other stuff using this (inside `select_qi`) which I haven't looked too thoroughly into.

Reply:
What I mean is, this is basically the same info as the keyframe set. i.e. this number changes each time the keyframe = true. Arguably it's still worth having as either you need to search for the last keyframe, or compare the current and last gop_input_frameno_start values. It's important that it stays in sync with the keyframes though. It would be nice if it was clearly set where keyframes are set.

Reply:
There's some non-obvious semantics on how and when it changes with respect to e.g. invalid frames. I'd wait for further simplification of `build_frame_properties` to fix this.

Reply:
OK